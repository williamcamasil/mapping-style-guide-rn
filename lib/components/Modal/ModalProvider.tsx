import React, {
  useState, ReactElement, useCallback, useMemo, useRef, Fragment, ComponentType, useContext, useEffect,
} from 'react';

import { ModalPropsType } from '.';

type ShowModalCallbackPropsType<P extends ModalPropsType > = Omit<P, 'visible' | 'isFocused' | 'modalKey' | 'theme'>;

export type ModalReturnType = {
  dismiss: () => void;
};

type ModalContextType = {
  showModal: (
    ModalComponent: ComponentType<ModalPropsType>,
    props?: ShowModalCallbackPropsType<ModalPropsType>
  ) => ModalReturnType;
  dismissAllModals: () => void;
};

const ModalContext = React.createContext<ModalContextType>({} as any);

type RenderedElementType = {
  modalKey: string;
  element: ReactElement<ModalPropsType>;
  notifyDismiss: () => void;
  notifyDismissEnd: () => void;
};

type ModalProviderPropsType = {
  children: React.ReactNode;
  /**
   * Tempo (em milissegundos) de tolerância entre a abertura de duas modais do mesmo tipo.
   *
   * Usado para evitar que o usuário abra várias vezes a mesma modal ao clicar
   * rapidamente no mesmo botão.
   *
   * Em ambiente de teste, você pode querer desativar esse comportamento, pois
   * em casos de testes os disparos de modais acontecem várias vezes repetidamente.
   *
   * Para desativar, basta passar zero.
   */
  throttleTimeout?: number;
};

const ModalProvider: React.FC<ModalProviderPropsType> = ({
  children, throttleTimeout = 500,
}) => {
  const lastModalShowRef = useRef(0);
  const lastElementTypeRef = useRef<ReactElement['type']>();
  const [renderedElements, setRenderedElements] = useState<RenderedElementType[]>([]);

  const lastRenderedElementsRef = useRef<RenderedElementType[]>([]);

  /**
   * Dispara os eventos de dismiss e dismissEnd de acordo com as mudanças
   * de estado das modais.
   */
  useEffect(() => {
    lastRenderedElementsRef.current.forEach(lastElement => {
      const element = renderedElements.find(el => el.modalKey === lastElement.modalKey);

      if (!element) {
        lastElement.notifyDismissEnd();
        return;
      }

      if (!element.element.props.visible && lastElement.element.props.visible) {
        lastElement.notifyDismiss();
      }
    });

    lastRenderedElementsRef.current = renderedElements;
  }, [renderedElements]);

  const currentRenderedElement = renderedElements[renderedElements.length - 1];

  const dismissAllModals = useCallback(() => {
    setRenderedElements(elements => (
      elements.map(element => {
        if (element.element.props.visible) {
          return {
            ...element,
            element: React.cloneElement(element.element, { visible: false }),
          };
        }

        return element;
      })
    ));
  }, []);

  const unmountModal = useCallback((modalKey: string) => {
    setRenderedElements(elements => {
      const elementIndex = elements.findIndex(element => element.modalKey === modalKey);

      const copyElements = [...elements];

      copyElements.splice(elementIndex, 1);

      return copyElements;
    });
  }, []);

  const dismissModal = useCallback((modalKey: string) => {
    setRenderedElements(elements => elements.map(element => {
      if (element.modalKey === modalKey && element.element.props.visible) {
        return {
          ...element,
          element: React.cloneElement(element.element, { visible: false }),
        };
      }

      return element;
    }));
  }, []);

  const shouldThrottle = useCallback((element: ReactElement) => {
    try {
      if (!throttleTimeout) return false;

      if (element.type !== lastElementTypeRef.current) return false;

      const timeSinceLastShow = Date.now() - lastModalShowRef.current;

      if (timeSinceLastShow > throttleTimeout) return false;

      return true;
    } finally {
      lastElementTypeRef.current = element.type;
      lastModalShowRef.current = Date.now();
    }
  }, [throttleTimeout]);

  const showModal = useCallback(<Props extends ModalPropsType>(
    ModalComponent: ComponentType<Props>,
    props?: ShowModalCallbackPropsType<Props>,
  ) => {
    const modalKey = `modal-element-${lastModalShowRef.current}${Date.now()}`;

    const result = {
      dismiss: () => dismissModal(modalKey),
    };

    setRenderedElements(elements => {
      // eslint-disable-next-line react-perf/jsx-no-new-function-as-prop
      const handleDismissModal = () => dismissModal(modalKey);

      // eslint-disable-next-line react-perf/jsx-no-new-function-as-prop
      const handleDismissEndModal = () => unmountModal(modalKey);

      const element = (
        <ModalComponent
          {...props as any}
          visible
          onDismiss={handleDismissModal}
          onDismissEnd={handleDismissEndModal}
        />
      );

      if (shouldThrottle(element)) {
        return elements;
      }

      const renderedElement: RenderedElementType = {
        modalKey,
        element,
        notifyDismiss: () => {
          props?.onDismiss?.();
        },
        notifyDismissEnd: () => {
          props?.onDismissEnd?.();
        },
      };

      return [...elements, renderedElement];
    });

    return result;
  }, [dismissModal, shouldThrottle, unmountModal]);

  const providerValue = useMemo(() => ({
    showModal,
    dismissAllModals,
  }), [
    showModal,
    dismissAllModals,
  ]);

  const renderModalElement = (item: RenderedElementType) => {
    const clonedElement = React.cloneElement(item.element, {
      isFocused: currentRenderedElement?.modalKey === item.modalKey,
      modalKey: item.modalKey,
    });

    return (
      <Fragment key={item.modalKey}>
        {clonedElement}
      </Fragment>
    );
  };

  return (
    <ModalContext.Provider value={providerValue}>
      {children}
      {renderedElements.map(renderModalElement)}
    </ModalContext.Provider>
  );
};

export type ShowModalType = <Props extends ModalPropsType>(
  ModalComponent: ComponentType<Props>,
  props?: ShowModalCallbackPropsType<Props>
) => ModalReturnType;

export function useModal() {
  const { showModal } = useContext(ModalContext);
  return showModal as ShowModalType;
}

export function useDismissAllModals() {
  const { dismissAllModals } = useContext(ModalContext);
  return dismissAllModals;
}

export default ModalProvider;
