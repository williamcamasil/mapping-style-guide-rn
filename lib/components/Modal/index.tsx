
import React, {
  useState, useMemo, useCallback, useRef, useEffect, PropsWithChildren,
} from 'react';
import {
  StyleSheet, Animated, BackHandler, Easing,
  SafeAreaView, View, StyleProp, ViewStyle,
  Platform,
} from 'react-native';

import { useDimensions, useViewStyles } from '../../hooks';

export type ModalPropsType = PropsWithChildren & {
  testID?: string;
  visible: boolean;
  isFocused: boolean;
  modalKey?: string;
  onDismiss?: () => any;
  dismissible?: boolean;
  children?: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  overlayStyle?: StyleProp<ViewStyle>;
  onDismissEnd?: () => any;
  withSafeAreaIOS?: boolean;
  skipStatusBarAndroid?: boolean;
  animationDuration?: number;
  analyticsModalName?: string;
};

const styles = StyleSheet.create({
  portal: {
    ...StyleSheet.absoluteFillObject,
    backgroundColor: 'transparent',
  },
  safeArea: {
    flex: 1,
  },
  overlay: {
    flex: 1,
  },
  modal: {
    flex: 1,
  },
});

const Modal = ({
  visible,
  onDismiss,
  dismissible = true,
  children,
  style,
  overlayStyle,
  onDismissEnd,
  withSafeAreaIOS = true,
  skipStatusBarAndroid = false,
  animationDuration = 250,
  isFocused,
  testID = undefined,
  analyticsModalName,
  modalKey,
}: ModalPropsType) => {
  const $visible = useRef(false);
  const [rendered, setRendered] = useState(visible);

  const { window, statusBar } = useDimensions();
  const windowHeight = Math.max(window.width, window.height);

  const overlayOpacity = useRef(new Animated.Value(0));

  const modalTranslateY = useMemo(
    () => new Animated.Value(windowHeight),
    [windowHeight],
  );

  const showModal = useCallback(() => {
    setRendered(true);
    Animated.parallel([
      Animated.timing(modalTranslateY, {
        toValue: 0,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
      Animated.timing(overlayOpacity.current, {
        toValue: 1,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
    ]).start();
  }, [animationDuration, modalTranslateY]);

  const hideModal = useCallback(() => {
    Animated.parallel([
      Animated.timing(modalTranslateY, {
        toValue: windowHeight,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
      Animated.timing(overlayOpacity.current, {
        toValue: 0,
        duration: animationDuration,
        easing: Easing.out(Easing.cubic),
        useNativeDriver: true,
      }),
    ]).start(() => {
      setRendered(false);
      onDismissEnd?.();
    });
  }, [modalTranslateY, windowHeight, animationDuration, overlayOpacity, onDismissEnd]);

  useEffect(() => {
    if ($visible.current === visible) return;

    $visible.current = visible;

    if (visible) {
      showModal();
    } else {
      hideModal();
    }
  }, [hideModal, showModal, visible]);

  useEffect(() => {
    if (visible) {
      const handleBack = () => {
        if (dismissible) {
          onDismiss?.();
        }
        return true;
      };

      const listener = BackHandler.addEventListener('hardwareBackPress', handleBack);

      return () => {
        listener.remove();
      };
    }
    return undefined;
  }, [visible, dismissible, onDismiss]);

  const overlayStyles = useViewStyles(() => [
    styles.overlay,
    overlayStyle,
    {
      opacity: overlayOpacity.current,
    },
  ], [overlayOpacity, overlayStyle]);

  const modalStyles = useViewStyles(() => [
    styles.modal,
    style,
    {
      marginTop: Platform.OS === 'android' && skipStatusBarAndroid ? statusBar.height : undefined,
      transform: [{
        translateY: modalTranslateY,
      }],
    },
  ], [modalTranslateY, skipStatusBarAndroid, statusBar.height, style]);

  if (!rendered) return null;

  const WrapperView = withSafeAreaIOS ? SafeAreaView : View;

  return (
    <View testID={testID} style={styles.portal} pointerEvents="box-none">
      <WrapperView testID="modal-safeArea" pointerEvents="box-none" style={styles.safeArea}>
        <Animated.View
          testID="modal-overlay"
          accessibilityViewIsModal
          accessibilityLiveRegion="polite"
          style={overlayStyles}
          onAccessibilityEscape={onDismiss}
        >
          <Animated.View
            testID="modal-content"
            style={modalStyles}
          >
            {children}
          </Animated.View>
        </Animated.View>
      </WrapperView>
    </View>
  );
};

export default Modal;
