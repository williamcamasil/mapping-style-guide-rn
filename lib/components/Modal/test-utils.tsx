import React, { useCallback } from 'react';
import { View } from 'react-native';

import { render } from '@testing-library/react-native';

import Modal, { ModalPropsType } from '.';
import { ThemeProvider } from '../../theme';
import Text from '../Text';
import ModalProvider, { ModalReturnType, useDismissAllModals, useModal } from './ModalProvider';

export const renderModal = async (modalProps: Partial<ModalPropsType>, throttleTimeout?: number) => {
  const ModalEmpty = (props: ModalPropsType) => (
    <Modal {...props}>
      <Text testID="modal-children-text">
        Sample text
      </Text>
    </Modal>
  );

  let showModal: (customProps?: Partial<ModalPropsType>) => ModalReturnType = () => ({
    dismiss: jest.fn(),
  });

  let dismissAllModals: () => void = jest.fn();

  const MyScreen = () => {
    const show = useModal();

    dismissAllModals = useDismissAllModals();

    showModal = useCallback((customProps?: Partial<ModalPropsType>) => show(ModalEmpty, {
      animationDuration: 0,
      ...modalProps,
      ...customProps,
    }), [show]);

    return <View testID="modal-screen" />;
  };

  const result = render(
    <ThemeProvider>
      <ModalProvider throttleTimeout={throttleTimeout}>
        <MyScreen />
      </ModalProvider>
    </ThemeProvider>,
  );

  await result.findByTestId('modal-screen');

  return { showModal, result, dismissAllModals };
};
