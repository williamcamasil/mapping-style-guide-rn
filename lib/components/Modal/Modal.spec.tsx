import React from 'react';
import { View, Text } from 'react-native';

import {
  render, act, waitFor,
} from '@testing-library/react-native';
import { delay } from 'mapping-context-rn';

import Modal from '.';
import { ModalReturnType } from './ModalProvider';
import { renderModal } from './test-utils';

describe('Modal snapshot', () => {
  it('should render before animation', () => {
    const tree = render(
      <Modal
        isFocused
        modalKey="abc"
        visible
        onDismiss={() => null}
        onDismissEnd={() => null}
        animationDuration={1000}
      >
        <View>
          <Text>Modal</Text>
        </View>
      </Modal>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render after animation', () => {
    const tree = render(
      <Modal
        isFocused
        modalKey="abc"
        visible
        onDismiss={() => null}
        onDismissEnd={() => null}
        animationDuration={0}
      >
        <View>
          <Text>Modal</Text>
        </View>
      </Modal>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render invisible', () => {
    const tree = render(
      <Modal
        isFocused={false}
        modalKey="abc"
        visible={false}
        animationDuration={0}
      >
        <View>
          <Text>Modal</Text>
        </View>
      </Modal>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render without iOS safe area', () => {
    const tree = render(
      <Modal
        isFocused
        modalKey="abc"
        visible
        withSafeAreaIOS={false}
        animationDuration={0}
      >
        <View>
          <Text>Modal</Text>
        </View>
      </Modal>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render with Android StatusBar skip', () => {
    const tree = render(
      <Modal
        isFocused
        modalKey="abc"
        visible
        skipStatusBarAndroid
        animationDuration={0}
      >
        <View>
          <Text>Modal</Text>
        </View>
      </Modal>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

});

describe('Modal events', () => {
  // it('should render with analyticsModalName', async () => {
  //   let spyScreenView: any;
  //   const screen = render(
  //     <AppProvider>
  //       <Modal
  //         isFocused
  //         visible
  //         analyticsModalName="sa_modal_example"
  //         animationDuration={0}
  //       >
  //         <View>
  //           <Text>Modal</Text>
  //         </View>
  //       </Modal>
  //     </AppProvider>,
  //   );

  //   await waitFor(() => screen.UNSAFE_getByType(Modal));

  //   await waitFor(() => expect(spyScreenView).toBeCalledWith('sa_modal_example'));

  //   const tree = screen.toJSON();

  //   expect(tree).toMatchSnapshot();
  // });
  it('should call onDismissEnd after dismiss', () => {
    const handleDismiss = jest.fn();

    const result = render(
      <Modal
        isFocused
        modalKey="abc"
        visible
        onDismissEnd={handleDismiss}
        animationDuration={0}
      />,
    );

    act(() => {
      result.update(
        <Modal
          isFocused
          modalKey="abc"
          visible={false}
          onDismissEnd={handleDismiss}
          animationDuration={0}
        />,
      );
    });

    expect(handleDismiss).toBeCalled();
  });

  it('should dismiss the modal with useDismissAllModals', async () => {
    const handleDismiss = jest.fn();

    const { showModal, dismissAllModals } = await renderModal({
      onDismiss: handleDismiss,
    });

    act(() => {
      showModal();
    });
    act(() => {
      dismissAllModals();
    });

    expect(handleDismiss).toBeCalled();
  });

  it('should ignore second show with throttle', async () => {
    const handleDismiss = jest.fn();

    const { showModal, result } = await renderModal({
      onDismiss: handleDismiss,
    });

    act(() => {
      showModal();
    });
    act(() => {
      showModal();
    });

    const children = result.getAllByTestId('modal-children-text');

    expect(children.length).toBe(1);
  });

  it('should show with throttle disabled', async () => {
    const handleDismiss = jest.fn();

    const { showModal, result } = await renderModal({
      onDismiss: handleDismiss,
    }, 0);

    act(() => {
      showModal();
    });
    act(() => {
      showModal();
    });

    const children = result.getAllByTestId('modal-children-text');

    expect(children.length).toBe(2);
  });

  it('should show with small throttle', async () => {
    const handleDismiss = jest.fn();

    const { showModal, result } = await renderModal({
      onDismiss: handleDismiss,
    }, 1);

    act(() => {
      showModal();
    });

    await delay(1);

    act(() => {
      showModal();
    });

    const children = result.getAllByTestId('modal-children-text');

    expect(children.length).toBe(2);
  });

  it('should track correct modal in render queue', async () => {
    const { showModal } = await renderModal({}, 0);

    act(() => {
      showModal({
        analyticsModalName: 'sa_first_modal',
      });
    });

    let secondModalReturn: ModalReturnType;
    act(() => {
      secondModalReturn = showModal({
        analyticsModalName: 'sa_second_modal',
      });
    });

    act(() => {
      secondModalReturn.dismiss();
    });
  });

  it('should not call dismiss more that once', async () => {
    const handleDismiss = jest.fn();

    const { showModal, result } = await renderModal({
      onDismiss: handleDismiss,
      animationDuration: 100,
    });

    let showResult: ModalReturnType;
    act(() => {
      showResult = showModal();
    });

    result.getByTestId('modal-children-text');

    act(() => {
      showResult.dismiss();
    });
    act(() => {
      showResult.dismiss();
    });

    await waitFor(() => {
      expect(result.queryByTestId('modal-children-text')).toBeNull();
    });

    expect(handleDismiss).toBeCalledTimes(1);

    act(() => {
      showResult.dismiss();
    });

    expect(handleDismiss).toBeCalledTimes(1);
  });

  it('should not call dismissAll more that once fot the same modal', async () => {
    const handleDismiss = jest.fn();

    const { showModal, result, dismissAllModals } = await renderModal({
      onDismiss: handleDismiss,
      animationDuration: 100,
    });

    act(() => {
      showModal();
    });

    result.getByTestId('modal-children-text');

    act(() => {
      dismissAllModals();
    });
    act(() => {
      dismissAllModals();
    });

    await waitFor(() => {
      expect(result.queryByTestId('modal-children-text')).toBeNull();
    });

    expect(handleDismiss).toBeCalledTimes(1);

    act(() => {
      dismissAllModals();
    });

    expect(handleDismiss).toBeCalledTimes(1);
  });
});
