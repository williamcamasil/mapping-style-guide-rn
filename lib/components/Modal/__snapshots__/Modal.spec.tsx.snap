// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal snapshot should render after animation 1`] = `
<View
  pointerEvents="box-none"
  style={
    {
      "backgroundColor": "transparent",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <RCTSafeAreaView
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
    testID="modal-safeArea"
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[Function]}
      style={
        {
          "flex": 1,
          "opacity": 1,
        }
      }
      testID="modal-overlay"
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "marginTop": undefined,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          }
        }
        testID="modal-content"
      >
        <View>
          <Text>
            Modal
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;

exports[`Modal snapshot should render before animation 1`] = `
<View
  pointerEvents="box-none"
  style={
    {
      "backgroundColor": "transparent",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <RCTSafeAreaView
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
    testID="modal-safeArea"
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      onAccessibilityEscape={[Function]}
      style={
        {
          "flex": 1,
          "opacity": 0,
        }
      }
      testID="modal-overlay"
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "marginTop": undefined,
            "transform": [
              {
                "translateY": 1334,
              },
            ],
          }
        }
        testID="modal-content"
      >
        <View>
          <Text>
            Modal
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;

exports[`Modal snapshot should render invisible 1`] = `null`;

exports[`Modal snapshot should render with Android StatusBar skip 1`] = `
<View
  pointerEvents="box-none"
  style={
    {
      "backgroundColor": "transparent",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <RCTSafeAreaView
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
    testID="modal-safeArea"
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      style={
        {
          "flex": 1,
          "opacity": 1,
        }
      }
      testID="modal-overlay"
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "marginTop": undefined,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          }
        }
        testID="modal-content"
      >
        <View>
          <Text>
            Modal
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;

exports[`Modal snapshot should render without iOS safe area 1`] = `
<View
  pointerEvents="box-none"
  style={
    {
      "backgroundColor": "transparent",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
    testID="modal-safeArea"
  >
    <View
      accessibilityLiveRegion="polite"
      accessibilityViewIsModal={true}
      collapsable={false}
      style={
        {
          "flex": 1,
          "opacity": 1,
        }
      }
      testID="modal-overlay"
    >
      <View
        collapsable={false}
        style={
          {
            "flex": 1,
            "marginTop": undefined,
            "transform": [
              {
                "translateY": 0,
              },
            ],
          }
        }
        testID="modal-content"
      >
        <View>
          <Text>
            Modal
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
